File: region/RegionLoader.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;

/*
 * Copyright (c) 2016-2017, Adam <Adam@sigterm.info>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
namespace OSRSCache.region
{
	using IndexType = OSRSCache.IndexType;
	using LocationsDefinition = OSRSCache.definitions.LocationsDefinition;
	using MapDefinition = OSRSCache.definitions.MapDefinition;
	using LocationsLoader = OSRSCache.definitions.loaders.LocationsLoader;
	using MapLoader = OSRSCache.definitions.loaders.MapLoader;
	using Archive = OSRSCache.fs.Archive;
	using Index = OSRSCache.fs.Index;
	using Storage = OSRSCache.fs.Storage;
	using Store = OSRSCache.fs.Store;
	using XteaKeyManager = OSRSCache.util.XteaKeyManager;


	public class RegionLoader
	{
		private const int MAX_REGION = 32768;

		private readonly Store store;
		private readonly Index index;
		private readonly XteaKeyManager keyManager;

		private readonly IDictionary<int, Region> regions = new Dictionary<int, Region>();
		private Region lowestX = null, lowestY = null;
		private Region highestX = null, highestY = null;

		public RegionLoader(Store store)
		{
			this.store = store;
			index = store.getIndex(IndexType.MAPS);
			keyManager = new XteaKeyManager();
			keyManager.loadKeys();
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public void loadRegions() throws java.io.IOException
		public virtual void loadRegions()
		{
			for (int i = 0; i < MAX_REGION; ++i)
			{
				Region region = this.loadRegionFromArchive(i);
				if (region != null)
				{
					regions[i] = region;
				}
			}
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public Region loadRegionFromArchive(int i) throws java.io.IOException
		public virtual Region loadRegionFromArchive(int i)
		{
			int x = i >> 8;
			int y = i & 0xFF;

			Storage storage = store.Storage;
			Archive map = index.findArchiveByName("m" + x + "_" + y);
			Archive land = index.findArchiveByName("l" + x + "_" + y);

			Debug.Assert((map == null) == (land == null));

			if (map == null || land == null)
			{
				return null;
			}

			sbyte[] data = map.decompress(storage.loadArchive(map));

			MapDefinition mapDef = (new MapLoader()).load(x, y, data);

			Region region = new Region(i);
			region.loadTerrain(mapDef);

			int[] keys = keyManager.getKeys(i);
			if (keys != null)
			{
				try
				{
					data = land.decompress(storage.loadArchive(land), keys);
					LocationsDefinition locDef = (new LocationsLoader()).load(x, y, data);
					region.loadLocations(locDef);
				}
				catch (IOException ex)
				{
					Console.WriteLine("Can't decrypt region " + i, ex);
				}
			}

			return region;
		}

		public virtual void calculateBounds()
		{
			foreach (Region region in regions.Values)
			{
				if (lowestX == null || region.BaseX < lowestX.BaseX)
				{
					lowestX = region;
				}

				if (highestX == null || region.BaseX > highestX.BaseX)
				{
					highestX = region;
				}

				if (lowestY == null || region.BaseY < lowestY.BaseY)
				{
					lowestY = region;
				}

				if (highestY == null || region.BaseY > highestY.BaseY)
				{
					highestY = region;
				}
			}
		}

		public virtual ICollection<Region> Regions
		{
			get
			{
				return regions.Values;
			}
		}

		public virtual Region findRegionForWorldCoordinates(int x, int y)
		{
			x = (int)((uint)x >> 6);
			y = (int)((uint)y >> 6);
			return regions[(x << 8) | y];
		}

		public virtual Region LowestX
		{
			get
			{
				return lowestX;
			}
		}

		public virtual Region LowestY
		{
			get
			{
				return lowestY;
			}
		}

		public virtual Region HighestX
		{
			get
			{
				return highestX;
			}
		}

		public virtual Region HighestY
		{
			get
			{
				return highestY;
			}
		}
	}

}

File: src/main/java/net/runelite/cache/region/RegionLoader.cpp
#include "RegionLoader.h"

namespace net::runelite::cache::region
{
	using IOException = java::io::IOException;
	using Collection = java::util::Collection;
	using HashMap = java::util::HashMap;
	using Map = java::util::Map;
	using Slf4j = lombok::extern_Keyword::slf4j::Slf4j;
	using IndexType = net::runelite::cache::IndexType;
	using LocationsDefinition = net::runelite::cache::definitions::LocationsDefinition;
	using MapDefinition = net::runelite::cache::definitions::MapDefinition;
	using LocationsLoader = net::runelite::cache::definitions::loaders::LocationsLoader;
	using MapLoader = net::runelite::cache::definitions::loaders::MapLoader;
	using Archive = net::runelite::cache::fs::Archive;
	using Index = net::runelite::cache::fs::Index;
	using Storage = net::runelite::cache::fs::Storage;
	using Store = net::runelite::cache::fs::Store;
	using KeyProvider = net::runelite::cache::util::KeyProvider;

	RegionLoader::RegionLoader(std::shared_ptr<Store> store, std::shared_ptr<KeyProvider> keyProvider) : store(store), index(store->getIndex(IndexType::MAPS)), keyProvider(keyProvider)
	{
	}

	void RegionLoader::loadRegions()
	{
		if (!this->regions.empty())
		{
			return;
		}

		for (int i = 0; i < MAX_REGION; ++i)
		{
			try
			{
				this->loadRegionFromArchive(i);
			}
			catch (const IOException &ex)
			{
				log::debug(L"Can't decrypt region " + std::to_wstring(i), ex);
			}
		}
	}

	std::shared_ptr<Region> RegionLoader::loadRegionFromArchive(int i)
	{
		int x = i >> 8;
		int y = i & 0xFF;

		std::shared_ptr<Storage> storage = store->getStorage();
		std::shared_ptr<Archive> map = index->findArchiveByName(L"m" + std::to_wstring(x) + L"_" + std::to_wstring(y));
		std::shared_ptr<Archive> land = index->findArchiveByName(L"l" + std::to_wstring(x) + L"_" + std::to_wstring(y));

		assert((map == nullptr) == (land == nullptr));

		if (map == nullptr || land == nullptr)
		{
			return nullptr;
		}

		std::vector<signed char> data = map->decompress(storage->loadArchive(map));

		std::shared_ptr<MapDefinition> mapDef = (std::make_shared<MapLoader>())->load(x, y, data);

		std::shared_ptr<Region> region = std::make_shared<Region>(i);
		region->loadTerrain(mapDef);

		std::vector<int> keys = keyProvider->getKey(i);
		if (!keys.empty())
		{
			data = land->decompress(storage->loadArchive(land), keys);
			std::shared_ptr<LocationsDefinition> locDef = (std::make_shared<LocationsLoader>())->load(x, y, data);
			region->loadLocations(locDef);
		}

		regions.emplace(i, region);

		return region;
	}

	std::shared_ptr<Region> RegionLoader::loadRegion(int id, std::shared_ptr<MapDefinition> map, std::shared_ptr<LocationsDefinition> locs)
	{
		std::shared_ptr<Region> r = std::make_shared<Region>(id);
		r->loadTerrain(map);
		if (locs != nullptr)
		{
			r->loadLocations(locs);
		}
		regions.emplace(id, r);
		return r;
	}

	void RegionLoader::calculateBounds()
	{
		for (auto region : regions)
		{
			if (lowestX == nullptr || region->second.getBaseX() < lowestX->getBaseX())
			{
				lowestX = region->second;
			}

			if (highestX == nullptr || region->second.getBaseX() > highestX->getBaseX())
			{
				highestX = region->second;
			}

			if (lowestY == nullptr || region->second.getBaseY() < lowestY->getBaseY())
			{
				lowestY = region->second;
			}

			if (highestY == nullptr || region->second.getBaseY() > highestY->getBaseY())
			{
				highestY = region->second;
			}
		}
	}

	std::shared_ptr<Collection<std::shared_ptr<Region>>> RegionLoader::getRegions()
	{
		return regions.values();
	}

	std::shared_ptr<Region> RegionLoader::findRegionForWorldCoordinates(int x, int y)
	{
		x = static_cast<int>(static_cast<unsigned int>(x) >> 6);
		y = static_cast<int>(static_cast<unsigned int>(y) >> 6);
		return regions[(x << 8) | y];
	}

	std::shared_ptr<Region> RegionLoader::findRegionForRegionCoordinates(int x, int y)
	{
		return regions[(x << 8) | y];
	}

	std::shared_ptr<Region> RegionLoader::getLowestX()
	{
		return lowestX;
	}

	std::shared_ptr<Region> RegionLoader::getLowestY()
	{
		return lowestY;
	}

	std::shared_ptr<Region> RegionLoader::getHighestX()
	{
		return highestX;
	}

	std::shared_ptr<Region> RegionLoader::getHighestY()
	{
		return highestY;
	}
}
